app/lib/fields.py

```diff
@@ -305,3 +308,180 @@ def update_choices(
         # Replace the field's attribute value
         self.choices = choices
         self.choices_updated = True
+
+
+class DateComponentField(BaseField):
+    """Handles day/month/year components and validates them as a complete date"""
+
+    def __init__(self, is_from_date=True, **kwargs):
+        super().__init__(**kwargs)
+        self.day = ""
+        self.month = ""
+        self.year = ""
+        self.is_from_date = is_from_date
+        self._form_data = None
+        self._components_extracted = False
+        self.was_expanded = False
+
+    def bind(self, name, value: list | str) -> None:
+        super().bind(name, "")
+        # Don't auto-detect is_from_date from field name in tests
+        # Let the constructor value take precedence
+
+    def set_form_data(self, form_data):
+        """Set form data and extract components immediately"""
+        self._form_data = form_data
+        if self.name and form_data:
+            self._extract_components(form_data, self.name)
+
+    def _extract_components(self, form_data, name):
+        """Extract day, month, year components from form data"""
+        if form_data:
+            self.day = str(form_data.get(f"{name}-day", "")).strip()
+            self.month = str(form_data.get(f"{name}-month", "")).strip()
+            self.year = str(form_data.get(f"{name}-year", "")).strip()
+            self._components_extracted = True
```

Comment:
same as above - consider to use form_kwargs in CatalogueSearchFormMixin. consider extracting the input data and build the value to bind.  This could be useful  https://github.com/nationalarchives/ds-catalogue/blob/243390422b7b6ef600882fd2a719915d895e6a34/app/lib/forms.py#L45-L50  as the form data appears over here.